### PR TITLE
[Bridge] Signed Arbitrary Message Pool

### PR DESCRIPTION
**Pull request by Kourin1996**
Fixes EDGE-42

# Description

This PR adds Signed Arbitrary Message Pool

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [ ] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [ ] I have tested this code manually

### Manual tests

Please complete this section if you ran manual tests for this functionality, otherwise delete it

# Documentation update

Please link the documentation update PR in this section if it's present, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: bridge/sam/pool.go [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+	"sync"
+	"sync/atomic"
+
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+type pool struct {
+	// write-lock is called only when changing validators process
+	// otherwise read-lock is called
+	// Changing validators will occur the most rarely (once per epoch)
```
Comment (zivkovicmilos) [inline bridge/sam/pool.go]: We can rephrase this to `write-lock is called only during the validator change process`.

Also, we can rephrase this part as well to `Changing validators occurs rarely (once per epoch)`

File: bridge/sam/pool.go [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+	// map from message Hash -> bool for message flags
+	consumedMap sync.Map
+	readyMap    sync.Map
+
+	// signatures
+	messageSignatures *messageSignaturesStore
+}
+
+// diffAddresses returns a list of the addresses that are in arr1 but not in arr2
+func diffAddresses(arr1, arr2 []types.Address) []types.Address {
```
Comment (zivkovicmilos) [inline bridge/sam/pool.go]: Nitpick, please consider changing the naming of these `arr1` and `arr2` 🙏

File: bridge/sam/pool.go [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+	defer p.changeValidatorsLock.RUnlock()
+
+	if p.hasConsumed(message.Hash) {
+		// we do no longer put the signature if the message has been consumed
+		return
+	}
+
+	p.setMessageBody(message.Hash, message.Body)
+	p.tryToPromote(message.Hash)
+}
```
Comment (zivkovicmilos) [inline bridge/sam/pool.go]: This block of code is exactly the same as `AddSignature`, apart from the `p.setMessageBody` here and `p.messageSignatures.Put` in `AddSignature`.

What do you think about extracting this common logic?

File: bridge/sam/pool.go [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+// Add adds new message with the signature to pool
+func (p *pool) AddSignature(signature *MessageSignature) {
+	p.changeValidatorsLock.RLock()
+	defer p.changeValidatorsLock.RUnlock()
+
+	if p.hasConsumed(signature.Hash) {
+		// we do no longer put the signature if the message has been consumed
```
Comment (zivkovicmilos) [inline bridge/sam/pool.go]: Typo here, should be `we no longer put the signature....`

File: bridge/sam/pool.go [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+		return
+	}
+
+	p.messageSignatures.Put(signature)
+	p.tryToPromote(signature.Hash)
+}
+
+// Consume sets the consumed flag and delete the message from pool
+func (p *pool) Consume(hash types.Hash) {
```
Comment (zivkovicmilos) [inline bridge/sam/pool.go]: Consume what?

This method name should be more descriptive

File: bridge/sam/pool.go [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+	p.consumedMap.Store(hash, true)
+
+	p.messageSignatures.RemoveMessage(hash)
+	p.messageMap.Delete(hash)
+	p.readyMap.Delete(hash)
+}
+
+// knows returns the flag indicating the message is known
+func (p *pool) knows(hash types.Hash) bool {
```
Comment (zivkovicmilos) [inline bridge/sam/pool.go]: `knows` is an abstract method name, please consider renaming it to something like `isMessageKnown`

File: bridge/sam/pool.go [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+}
+
+// UpdateValidatorSet update validators and threshold
+// This process blocks other processes because messages would lose the signatures
+func (p *pool) UpdateValidatorSet(validators []types.Address, threshold uint64) {
+	p.changeValidatorsLock.Lock()
+	defer p.changeValidatorsLock.Unlock()
+
+	oldValidators := p.validators
+	oldThreshold := p.threshold //nolint
```
Comment (zivkovicmilos) [inline bridge/sam/pool.go]: Why the `//nolint`?

File: bridge/sam/pool.go [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+type pool struct {
+	// write-lock is called only when changing validators process
+	// otherwise read-lock is called
+	// Changing validators will occur the most rarely (once per epoch)
+	changeValidatorsLock sync.RWMutex
+	validators           []types.Address
+	threshold            uint64 // required number of signatures for ready
```
Comment (zivkovicmilos) [inline bridge/sam/pool.go]: Please consider renaming this to `signatureThreshold`, as it's not immediately clear from the usage what it represents

File: bridge/sam/pool.go [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+	diff := make([]types.Address, 0)
+
+	for _, addr := range arr1 {
+		if !arr2Map[addr] {
+			diff = append(diff, addr)
+		}
+	}
+
+	return diff
+}
```
Comment (zivkovicmilos) [inline bridge/sam/pool.go]: What do you think about extracting this method to the helper / common package?

File: bridge/sam/pool.go [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+
+// canDemote return the flag indicating it's possible to change status to pending
+func (p *pool) canDemote(hash types.Hash) bool {
+	return !p.canPromote(hash)
+}
+
+// tryToPromote checks the number of signatures and threshold and update message status to ready if need
+func (p *pool) tryToPromote(hash types.Hash) {
+	fmt.Printf("tryToPromote %v\n", hash)
```
Comment (zivkovicmilos) [inline bridge/sam/pool.go]: Leftover print

File: bridge/sam/pool.go [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+	if newNumSignatures < 0 {
+		return 0
+	}
+
+	return uint64(newNumSignatures)
+}
+
+// IncrementNumSignatures decrements SignatureCount and return new count
+func (e *signedMessageEntry) DecrementNumSignatures() uint64 {
+	newNumSignatures := atomic.AddInt64(&e.SignatureCount, ^int64(0))
```
Comment (zivkovicmilos) [inline bridge/sam/pool.go]: Why decrement with `^int64(0)` instead of just `-1`?

File: bridge/sam/pool.go [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+
+	return uint64(count)
+}
+
+// IncrementNumSignatures increments SignatureCount and return new count
+func (e *signedMessageEntry) IncrementNumSignatures() uint64 {
+	newNumSignatures := atomic.AddInt64(&e.SignatureCount, 1)
+	if newNumSignatures < 0 {
+		return 0
+	}
```
Comment (zivkovicmilos) [inline bridge/sam/pool.go]: Is this even possible?

**Review by zivkovicmilos — APPROVED** [review]
Looks good to me 💯 

I've left some comments you need to address.
Please resolve the linting errors 🙏 

I wanted to raise a discussion regarding persistence in this module, as far as I can see we keep these values in memory and don't persist them anywhere. Do you think we even need to bother with persisting the info stored in the SAM pool, as the messages themselves are temporary?

cc @lazartravica

File: bridge/sam/pool.go (hunk 7 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Actually it's necessary. It throws error because this declaration is far from for-loop which use this variable. But need to set here

File: bridge/sam/pool.go (hunk 2 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: `minuend` and `subtrahend`? They are not clear

File: bridge/sam/pool.go (hunk 3 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: I don't think we can extract common logic.

File: bridge/sam/pool.go (hunk 1 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Fixed at c92a67bf10bc73bcdf8929dca389a8e53020bf04

File: bridge/sam/pool.go (hunk 4 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Fixed at c92a67bf10bc73bcdf8929dca389a8e53020bf04

File: bridge/sam/pool.go (hunk 5 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Fixed at c92a67bf10bc73bcdf8929dca389a8e53020bf04

File: bridge/sam/pool.go (hunk 6 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Fixed at c92a67bf10bc73bcdf8929dca389a8e53020bf04

File: bridge/sam/pool.go (hunk 8 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Fixed at c92a67bf10bc73bcdf8929dca389a8e53020bf04

File: bridge/sam/pool.go (hunk 9 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Moved at c92a67bf10bc73bcdf8929dca389a8e53020bf04

File: bridge/sam/pool.go (hunk 10 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Removed at c92a67bf10bc73bcdf8929dca389a8e53020bf04

File: bridge/sam/pool.go (hunk 11 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Fixed at c92a67bf10bc73bcdf8929dca389a8e53020bf04

File: bridge/sam/pool.go (hunk 12 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Fixed at c92a67bf10bc73bcdf8929dca389a8e53020bf04

File: bridge/sam/pool.go (hunk 2 above)
Comment (zivkovicmilos) [reply bridge/sam/pool.go]: Something like `setA` and `setB`

File: bridge/sam/pool.go (hunk 3 above)
Comment (zivkovicmilos) [reply bridge/sam/pool.go]: Why not?

This entire block of code is identical:
```go
p.changeValidatorsLock.RLock()
	defer p.changeValidatorsLock.RUnlock()

	if p.hasConsumed(message.Hash) {
		// we do no longer put the signature if the message has been consumed
		return
	}
```

File: bridge/sam/pool.go (hunk 7 above)
Comment (zivkovicmilos) [reply bridge/sam/pool.go]: Aha, got it.

What's the linter that fails? if-short?

File: bridge/sam/pool.go (hunk 3 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: p.changeValidatorsLock.RUnlock() will be called the end of `AddMessage` and `AddSignature`. Should I create function that takes callback function? I guess it makes the flow unclear. I thought didn't need to save a few lines

File: bridge/sam/pool.go (hunk 7 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: ifshort

File: bridge/sam/pool.go (hunk 3 above)
Comment (zivkovicmilos) [reply bridge/sam/pool.go]: Got it, let's not bundle it, it's not breaking 👍

File: bridge/sam/pool.go (hunk 7 above)
Comment (zivkovicmilos) [reply bridge/sam/pool.go]: It's some bug with the ifshort linter itself, you can then change the `nolint` to `nolint:ifshort` 👍

File: bridge/sam/pool.go [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+		for _, hash := range maybeDemotableHashes {
+			p.tryToDemote(hash)
+		}
+	}
+}
+
+// canPromote return the flag indicating it's possible to change status to ready
+// message need to have enough signatures and be known by pool for promotion
+func (p *pool) canPromote(hash types.Hash) bool {
+	isKnown := p.isMessageKnown(hash)
```
Comment (lazartravica) [inline bridge/sam/pool.go:160]: Why do we need to check both whether the message is known and if it has enough signatures?

Wouldn't it be enough just to check the signatures?

**Review by lazartravica — CHANGES_REQUESTED** [review]
LGTM! Left one minor comment I would appreciate clarification on :)

File: bridge/sam/pool.go (hunk 13 above)
Comment (Kourin1996) [reply bridge/sam/pool.go:160]: To pick events up, you need to have data and enough signatures. Because expecting other validators to send hash and its signature by gossip. When validating, you can just check signature count.

File: bridge/sam/pool.go [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+
+// knows returns the flag indicating the message is known
+func (p *pool) isMessageKnown(hash types.Hash) bool {
+	_, ok := p.messageMap.Load(hash)
+	if !ok {
+		return false
+	}
+
+	return ok
+}
```
Comment (dbrajovic) [inline bridge/sam/pool.go]: Why not just return `ok` ?

File: bridge/sam/pool.go [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+	}
+
+	return body
+}
+
+// signedMessageEntry is representing the data stored in messageSignaturesStore
+type signedMessageEntry struct {
+	Hash           types.Hash
+	Signatures     sync.Map
+	SignatureCount int64
```
Comment (dbrajovic) [inline bridge/sam/pool.go:243]: Perhaps this field can be of type `uint64` - this would remove the `uint64()` casting of the return values in `NumSignatures`, `IncrementNumSignatures` and `DecrementNumSignatures`

File: bridge/sam/pool.go (hunk 14 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Good catch. Fixed at 95405e5209abfed5ccbad58db15f2550f6d04bf5

File: bridge/sam/pool.go (hunk 7 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Added 95405e5209abfed5ccbad58db15f2550f6d04bf5

File: bridge/sam/pool.go (hunk 2 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: To be honest, no big difference. But I renamed 95405e5209abfed5ccbad58db15f2550f6d04bf5

**Review by zivkovicmilos — APPROVED** [review]

File: bridge/sam/pool.go (hunk 13 above)
Comment (lazartravica) [reply bridge/sam/pool.go:160]: Thank you for the clarification, resolving this comment @Kourin1996

File: bridge/sam/pool_test.go [hunk 16]
```diff
@@ -0,0 +1,617 @@
...
+		threshold      uint64
+		knownMessages  []Message
+		consumedHashes []types.Hash
+		// inputs
+		signatures []MessageSignature
+		// outputs
+		numReadyMessages int
+	}{
+		{
+			name: "shouldn't promote message if the message doesn't have enough message",
```
Comment (0xAleksaOpacic) [inline bridge/sam/pool_test.go]: Typo 
....doesn't have enough signatures *

File: bridge/sam/pool.go [hunk 17]
```diff
@@ -0,0 +1,372 @@
...
+	p.changeValidatorsLock.Lock()
+	defer p.changeValidatorsLock.Unlock()
+
+	oldValidators := p.validators
+	oldThreshold := p.signatureThreshold //nolint:nolintlint //nolint:ifshort
+
+	p.validators = validators
+	atomic.StoreUint64(&p.signatureThreshold, threshold)
+
+	var maybeDemotableHashes []types.Hash
```
Comment (0xAleksaOpacic) [inline bridge/sam/pool.go]: typo
Demotable -> Demoted

**Review by 0xAleksaOpacic — APPROVED** [review]
LGTM!  Thank you on a great code coverage.

File: bridge/sam/pool_test.go (hunk 16 above)
Comment (Kourin1996) [reply bridge/sam/pool_test.go]: Thanks. I fixed at 66dd51ca6a17686e116cf8a384877f4fce953dee

File: bridge/sam/pool.go (hunk 17 above)
Comment (Kourin1996) [reply bridge/sam/pool.go]: Thanks. I fixed at 66dd51ca6a17686e116cf8a384877f4fce953dee

**Review by 0xAleksaOpacic — APPROVED** [review]

**Review by dbrajovic — APPROVED** [review]
